examples/nametoken/src/lib.rs

```diff
@@ -0,0 +1,1849 @@
+    }
+
+    /// Renew an existing name
+    fn renew(&mut self, name: &ContractTokenId) -> ContractResult<()> {
+        let mut entry = self
+            .all_names
+            .entry(*name)
+            .occupied_or(ContractError::Custom(CustomContractError::NameNotFound))?;
+        let new_expires = entry
+            .get_ref()
+            .name_expires
+            .checked_add(Duration::from_days(REGISTRATION_PERIOD_DAYS))
+            .ok_or(ContractError::Custom(CustomContractError::InvokeContractError))?;
+        entry.modify(|x| x.name_expires = new_expires);
+        Ok(())
+    }
+
+    /// Check that the token ID currently exists in this contract.
+    #[inline(always)]
+    fn contains_token(&self, token_id: &ContractTokenId) -> bool {
+        self.all_names.get(token_id).is_some()
+    }
+
+    /// Get the current balance of a given token ID for a given address.
+    /// Results in an error if the token ID does not exist in the state. Since
+    /// this contract only contains NFTs, the balance will always be either 1
+    /// or 0.
+    fn balance(
+        &self,
+        now: Timestamp,
+        token_id: &ContractTokenId,
+        address: &Address,
+    ) -> ContractResult<ContractTokenAmount> {
+        ensure!(self.contains_token(token_id), ContractError::InvalidTokenId);
+        let name_info = self.all_names.get(token_id).ok_or(ContractError::InvalidTokenId)?;
+        let balance = if let Address::Account(addr) = address {
+            self.state
+                .get(addr)
+                .map(|address_state| {
+                    // tokens are only considered owned by an address if they are not expired
+                    if address_state.owned_names.contains(token_id) && now <= name_info.name_expires
+                    {
+                        1
+                    } else {
+                        0
+                    }
+                })
+                .unwrap_or(0)
+        } else {
+            0
+        };
+        Ok(balance.into())
+    }
+
+    /// Check if a given address is an operator of a given owner address.
+    fn is_operator(&self, address: &Address, owner: &AccountAddress) -> bool {
+        self.state
+            .get(owner)
+            .map(|address_state| address_state.operators.contains(address))
+            .unwrap_or(false)
+    }
+
+    /// Update the state with a transfer of some token.
+    /// Results in an error if the token ID does not exist in the state or if
+    /// the from address have insufficient tokens to do the transfer.
+    fn transfer(
+        &mut self,
+        token_id: &ContractTokenId,
+        amount: ContractTokenAmount,
+        from: &Address,
+        to: &Address,
+        state_builder: &mut StateBuilder<S>,
+    ) -> ContractResult<()> {
+        ensure!(self.contains_token(token_id), ContractError::InvalidTokenId);
+        // A zero transfer does not modify the state.
+        if amount == 0.into() {
+            return Ok(());
+        }
```

Comment:
You should check the `from` and `to` address before this point, since a zero transfer between contracts would succeed right now.
I know it is nitpicking, but CIS-2 have a requirement that zero transfers must be treated as a normal transfer (seen from outside of the contract)

Reply:
Yes, fair enough; fixed.